WalletWasabi/Transactions/TransactionBuilding/SmartCoinSelector.cs

```diff
@@ -29,23 +30,54 @@ public IEnumerable<ICoin> Select(IEnumerable<ICoin> coins, IMoney target)
 				throw new InsufficientBalanceException(targetMoney, available);
 			}
 
-			// Group coins by scriptPubKey in order to treat them all as one.
-			var coinsByScriptPubKey = UnspentCoins
+			var uniqueClusters = UnspentCoins
+				.Select(coin => coin.Clusters)
+				.Distinct();
+
+			var clusters = (uniqueClusters.Count() < 10)
+				? uniqueClusters
+					.CombinationsWithoutRepetition(ofLength: 1, upToLength: 6)
+					.Select(clusterCombination => UnspentCoins.Where(coin => clusterCombination.Contains(coin.Clusters)))
+					.ToList()
+				: new List<IEnumerable<SmartCoin>>();
+
+			clusters.Add(UnspentCoins);
+
+			var coinsByCluster = clusters
+				.Select(coins => (Coins: coins, Privacy: 1.0m / coins.SelectMany(x => x.Clusters.KnownBy).Count()))
+				.Select(group => new
+				{
+					group.Coins,
+					Unconfirmed = group.Coins.Any(x => !x.Confirmed),    // If group has an unconfirmed, then the whole group is unconfirmed.
+					AnonymitySet = group.Coins.Min(x => x.AnonymitySet), // The group is as anonymous as its weakest member.
+					ClusterPrivacy = group.Privacy, // The number people/entities that know the cluster.
+					Amount = group.Coins.Sum(x => x.Amount)
+				});
+
+			var coinsInBestCluster = coinsByCluster
+				.Where(group => group.Amount >= targetMoney)
+				.OrderBy(group => group.Unconfirmed)
+				.ThenByDescending(group => group.AnonymitySet)     // Always try to spend/merge the largest anonset coins first.
+				.ThenByDescending(group => group.ClusterPrivacy)   // Select lesser-known coins.
+				.ThenByDescending(group => group.Amount)           // Then always try to spend by amount.
+				.First()
+				.Coins;
+
+			var coinsInBestClusterByScript = coinsInBestCluster
 				.GroupBy(c => c.ScriptPubKey)
 				.Select(group => new
 				{
 					Coins = group,
 					Unconfirmed = group.Any(x => !x.Confirmed),    // If group has an unconfirmed, then the whole group is unconfirmed.
```

Comment:
Is this used later?

Reply:
Yes, it is used. I am not sure if i am getting the question right.

Reply:
I replaced the code with this and everything still works. Even the tests are passing. 

```C#
			var coinsInBestClusterByScript = coinsInBestCluster
				.GroupBy(c => c.ScriptPubKey)
				.Select(group => new
				{
					Coins = group
				});
```

Reply:
Excellent point! The thing is we don't have a test to prove that code is needed. Everything works because coins with the same `ScriptPubKey` belong always to the same cluster and then they have the same cluster privacy so, when we sorted the coins they usually come one next to the other, that makes it highly probable to select them together. However, that is not guaranteed, the `coinsInBestClusterByScript` makes sure they are always added together. 

Anyway, I will review it because I think it can be improved a bit. Thanks @molnard.

Reply:
Ohh, I misunderstood your comment. Now I get it. You are right. Let me fix it.

Reply:
Done!